<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Subsetting PyRanges</title>
  <meta name="description" content="Subsetting PyRanges">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Subsetting PyRanges" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Subsetting PyRanges" />
  
  
  

<meta name="author" content="Endre Bakken Stovner">


<meta name="date" content="2022-01-11">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="printing-pyranges.html">
<link rel="next" href="sorting-pyranges.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>6</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="7" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>7</b> Piping PyRanges</a></li>
<li class="chapter" data-level="8" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>8</b> Printing PyRanges</a></li>
<li class="chapter" data-level="9" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>9</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="10" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>10</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="11" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>11</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="12" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>12</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="13" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>13</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="14" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>14</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="15" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>15</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="16" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>16</b> Joining Ranges</a></li>
<li class="chapter" data-level="17" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>17</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="18" data-path="finding-the-k-closest-intervals.html"><a href="finding-the-k-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the k closest intervals</a></li>
<li class="chapter" data-level="19" data-path="create-count-matrix-from-multiple-pyranges.html"><a href="create-count-matrix-from-multiple-pyranges.html"><i class="fa fa-check"></i><b>19</b> Create count-matrix from multiple PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-colocalization-and-co-occurence-measures.html"><a href="statistics-colocalization-and-co-occurence-measures.html"><i class="fa fa-check"></i><b>20</b> Statistics: colocalization and co-occurence measures</a></li>
<li class="chapter" data-level="21" data-path="statistics-simes-method.html"><a href="statistics-simes-method.html"><i class="fa fa-check"></i><b>21</b> Statistics: Simes method</a></li>
<li class="chapter" data-level="22" data-path="statistics-compute-fisher-exact.html"><a href="statistics-compute-fisher-exact.html"><i class="fa fa-check"></i><b>22</b> Statistics: compute Fisher exact</a></li>
<li class="chapter" data-level="23" data-path="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><a href="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><i class="fa fa-check"></i><b>23</b> Statistics: computing the Matthews correlation coeficcient between ranges</a></li>
<li class="chapter" data-level="24" data-path="statistics-rowbased-operations.html"><a href="statistics-rowbased-operations.html"><i class="fa fa-check"></i><b>24</b> Statistics: rowbased operations</a></li>
<li class="chapter" data-level="25" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>25</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="26" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>26</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="27" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>27</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="28" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>28</b> Subsetting Rles</a></li>
<li class="chapter" data-level="29" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>29</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="30" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>30</b> Using multiple cores</a></li>
<li class="chapter" data-level="31" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>31</b> GenomicFeatures: methods using genomic context</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Subsetting PyRanges</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="iterating-over-pyranges" class="section level1" number="9">
<h1><span class="header-section-number">9</span> Iterating over PyRanges</h1>
<p>PyRanges can be iterated over by a simple for loop, and the data is guaranteed
to come back in natsorted order:</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb102-1"><a href="iterating-over-pyranges.html#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb102-2"><a href="iterating-over-pyranges.html#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb102-3"><a href="iterating-over-pyranges.html#cb102-3" aria-hidden="true" tabindex="-1"></a>exons <span class="op">=</span> pr.data.exons()</span>
<span id="cb102-4"><a href="iterating-over-pyranges.html#cb102-4" aria-hidden="true" tabindex="-1"></a>cpg <span class="op">=</span> pr.data.cpg()</span>
<span id="cb102-5"><a href="iterating-over-pyranges.html#cb102-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k, df <span class="kw">in</span> cpg:</span>
<span id="cb102-6"><a href="iterating-over-pyranges.html#cb102-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb102-7"><a href="iterating-over-pyranges.html#cb102-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(df.head(<span class="dv">3</span>))</span></code></pre></div>
<pre><code>## chrX
##   Chromosome   Start     End  CpG
## 0       chrX   64181   64793   62
## 1       chrX   69133   70029  100
## 2       chrX  148685  149461   85
## chrY
##     Chromosome  Start    End  CpG
## 896       chrY  14181  14793   62
## 897       chrY  19133  20029  100
## 898       chrY  98685  99461   85</code></pre>
<p>If the data is unstranded, the key is just the chromosome names, but if the data
is stranded the key is a tuple of the chromosome and strand:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb104-1"><a href="iterating-over-pyranges.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k, df <span class="kw">in</span> exons:</span>
<span id="cb104-2"><a href="iterating-over-pyranges.html#cb104-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb104-3"><a href="iterating-over-pyranges.html#cb104-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(df.head(<span class="dv">3</span>))</span></code></pre></div>
<pre><code>## (&#39;chrX&#39;, &#39;+&#39;)
##   Chromosome      Start  ...  Score Strand
## 0       chrX  135721701  ...      0      +
## 2       chrX  135574120  ...      0      +
## 3       chrX   47868945  ...      0      +
## 
## [3 rows x 6 columns]
## (&#39;chrX&#39;, &#39;-&#39;)
##   Chromosome      Start  ...  Score Strand
## 1       chrX   49069126  ...      0      -
## 4       chrX  154006958  ...      0      -
## 6       chrX   52257919  ...      0      -
## 
## [3 rows x 6 columns]
## (&#39;chrY&#39;, &#39;+&#39;)
##    Chromosome     Start  ...  Score Strand
## 14       chrY   1693161  ...      0      +
## 82       chrY   1357411  ...      0      +
## 98       chrY  59233166  ...      0      +
## 
## [3 rows x 6 columns]
## (&#39;chrY&#39;, &#39;-&#39;)
##   Chromosome     Start  ...  Score Strand
## 5       chrY   1481624  ...      0      -
## 8       chrY  15526614  ...      0      -
## 9       chrY  15591393  ...      0      -
## 
## [3 rows x 6 columns]</code></pre>
<p>If you would like to iterate over the chromosomes in a stranded PyRange the idiom is</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb106-1"><a href="iterating-over-pyranges.html#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> c <span class="kw">in</span> exons.chromosomes:</span>
<span id="cb106-2"><a href="iterating-over-pyranges.html#cb106-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(c)</span>
<span id="cb106-3"><a href="iterating-over-pyranges.html#cb106-3" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> exons[c].df</span>
<span id="cb106-4"><a href="iterating-over-pyranges.html#cb106-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(df.head())</span></code></pre></div>
<pre><code>## chrX
##   Chromosome      Start  ...  Score Strand
## 0       chrX  135721701  ...      0      +
## 1       chrX  135574120  ...      0      +
## 2       chrX   47868945  ...      0      +
## 3       chrX   77294333  ...      0      +
## 4       chrX   91090459  ...      0      +
## 
## [5 rows x 6 columns]
## chrY
##   Chromosome     Start  ...  Score Strand
## 0       chrY   1693161  ...      0      +
## 1       chrY   1357411  ...      0      +
## 2       chrY  59233166  ...      0      +
## 3       chrY   1693161  ...      0      +
## 4       chrY   1664276  ...      0      +
## 
## [5 rows x 6 columns]</code></pre>
<p>but notice that we need the .df accessor, because subsetting a PyRange always
returns a PyRange.</p>
<p>There are three more ways to iterate over a pyrange, namely the keys, values and
items methods. These return a list, not a custom object like the python dict.</p>
<p>If you want to create a new PyRange while iterating over it, the idiom is</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb108-1"><a href="iterating-over-pyranges.html#cb108-1" aria-hidden="true" tabindex="-1"></a>d <span class="op">=</span> {}</span>
<span id="cb108-2"><a href="iterating-over-pyranges.html#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k, df <span class="kw">in</span> exons:</span>
<span id="cb108-3"><a href="iterating-over-pyranges.html#cb108-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(k)</span>
<span id="cb108-4"><a href="iterating-over-pyranges.html#cb108-4" aria-hidden="true" tabindex="-1"></a>    d[k] <span class="op">=</span> df.head(<span class="dv">3</span>)</span></code></pre></div>
<pre><code>## (&#39;chrX&#39;, &#39;+&#39;)
## (&#39;chrX&#39;, &#39;-&#39;)
## (&#39;chrY&#39;, &#39;+&#39;)
## (&#39;chrY&#39;, &#39;-&#39;)</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb110-1"><a href="iterating-over-pyranges.html#cb110-1" aria-hidden="true" tabindex="-1"></a>new_gr <span class="op">=</span> pr.PyRanges(d)</span>
<span id="cb110-2"><a href="iterating-over-pyranges.html#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(new_gr)</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+-------+
## | Chromosome   | Start     | End       | +3    |
## | (category)   | (int32)   | (int32)   | ...   |
## |--------------+-----------+-----------+-------|
## | chrX         | 135721701 | 135721963 | ...   |
## | chrX         | 135574120 | 135574598 | ...   |
## | chrX         | 47868945  | 47869126  | ...   |
## | chrX         | 49069126  | 49069255  | ...   |
## | ...          | ...       | ...       | ...   |
## | chrY         | 59233166  | 59233257  | ...   |
## | chrY         | 1481624   | 1481747   | ...   |
## | chrY         | 15526614  | 15526673  | ...   |
## | chrY         | 15591393  | 15592550  | ...   |
## +--------------+-----------+-----------+-------+
## Stranded PyRanges object has 12 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 3 hidden columns: Name, Score, Strand</code></pre>
<p>Note that this is basically the same as using the apply method of the PyRange,
only that the for loops are never multithreaded.</p>
<p>To iterate over multiple pyranges you can use pr.itergrs. It iterates
over multiple pyranges at the same time, returning the dfs belonging to the same
Chromosome or Chromosome/Strand at each iteration. Missing entries in one or
more PyRanges returns empty dfs.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb112-1"><a href="iterating-over-pyranges.html#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb112-2"><a href="iterating-over-pyranges.html#cb112-2" aria-hidden="true" tabindex="-1"></a>l <span class="op">=</span> pr.random(<span class="dv">25</span>), pr.random(<span class="dv">25</span>), pr.random(<span class="dv">25</span>)</span>
<span id="cb112-3"><a href="iterating-over-pyranges.html#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> key, grs <span class="kw">in</span> pr.itergrs(l, keys<span class="op">=</span><span class="va">True</span>, strand<span class="op">=</span><span class="va">True</span>):</span>
<span id="cb112-4"><a href="iterating-over-pyranges.html#cb112-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(key)</span>
<span id="cb112-5"><a href="iterating-over-pyranges.html#cb112-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(grs)</span></code></pre></div>
<pre><code>## (&#39;chr1&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 9       chr1  78901883  78901983      +,    Chromosome     Start       End Strand
## 15       chr1  77123011  77123111      +,    Chromosome     Start       End Strand
## 14       chr1  43313216  43313316      +
## 16       chr1  21537656  21537756      +]
## (&#39;chr1&#39;, &#39;-&#39;)
## [   Chromosome     Start       End Strand
## 10       chr1  91682232  91682332      -
## 11       chr1  22390062  22390162      -,    Chromosome     Start       End Strand
## 14       chr1  25415878  25415978      -,    Chromosome      Start        End Strand
## 15       chr1  139586949  139587049      -
## 17       chr1  163707001  163707101      -]
## (&#39;chr2&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 11       chr2  71413643  71413743      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr3&#39;, &#39;-&#39;)
## [   Chromosome      Start        End Strand
## 20       chr3   23220923   23221023      -
## 21       chr3  171874976  171875076      -
## 22       chr3  127345230  127345330      -,    Chromosome      Start        End Strand
## 16       chr3   38918800   38918900      -
## 17       chr3   48958106   48958206      -
## 18       chr3  174381532  174381632      -,   Chromosome      Start        End Strand
## 4       chr3  182725970  182726070      -
## 5       chr3  114128060  114128160      -
## 6       chr3   76465820   76465920      -]
## (&#39;chr4&#39;, &#39;+&#39;)
## [  Chromosome      Start        End Strand
## 2       chr4  121184930  121185030      +
## 3       chr4  135694380  135694480      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 8       chr4  176990919  176991019      +]
## (&#39;chr4&#39;, &#39;-&#39;)
## [  Chromosome      Start        End Strand
## 1       chr4  128407932  128408032      -,    Chromosome     Start       End Strand
## 24       chr4  55012884  55012984      -,   Chromosome      Start        End Strand
## 7       chr4  170602957  170603057      -]
## (&#39;chr5&#39;, &#39;+&#39;)
## [  Chromosome      Start        End Strand
## 8       chr5  161673901  161674001      +,   Chromosome    Start      End Strand
## 4       chr5  3068940  3069040      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr5&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 5       chr5  88227301  88227401      -,    Chromosome      Start        End Strand
## 21       chr5  116761061  116761161      -]
## (&#39;chr6&#39;, &#39;+&#39;)
## [   Chromosome     Start       End Strand
## 13       chr6  54038683  54038783      +
## 14       chr6  95965085  95965185      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr6&#39;, &#39;-&#39;)
## [   Chromosome      Start        End Strand
## 15       chr6  127849728  127849828      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 18       chr6  95470497  95470597      -]
## (&#39;chr7&#39;, &#39;+&#39;)
## [   Chromosome      Start        End Strand
## 23       chr7  157719940  157720040      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr7&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome      Start        End Strand
## 22       chr7  114925887  114925987      -]
## (&#39;chr8&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 2       chr8  116575415  116575515      +
## 3       chr8   68539597   68539697      +]
## (&#39;chr9&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome      Start        End Strand
## 12       chr9  116842217  116842317      +
## 13       chr9  123009191  123009291      +,    Chromosome      Start        End Strand
## 13       chr9  129996610  129996710      +]
## (&#39;chr9&#39;, &#39;-&#39;)
## [   Chromosome     Start       End Strand
## 18       chr9  42559143  42559243      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr10&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1      chr10  46365987  46366087      +]
## (&#39;chr10&#39;, &#39;-&#39;)
## [   Chromosome     Start       End Strand
## 12      chr10  77866806  77866906      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr10  24819870  24819970      -]
## (&#39;chr11&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 2      chr11   12149988   12150088      +
## 3      chr11  111013598  111013698      +,    Chromosome     Start       End Strand
## 19      chr11  37776736  37776836      +]
## (&#39;chr11&#39;, &#39;-&#39;)
## [   Chromosome      Start        End Strand
## 19      chr11  111814539  111814639      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr12&#39;, &#39;+&#39;)
## [  Chromosome      Start        End Strand
## 4      chr12   18289263   18289363      +
## 5      chr12  108211715  108211815      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 24      chr12  16744339  16744439      +]
## (&#39;chr13&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 8       chr13  69912333  69912433      +
## 10      chr13  50907161  50907261      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr13&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 9      chr13  21529544  21529644      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr15&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 6      chr15  17287546  17287646      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr15&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 7      chr15  20421957  20422057      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr16&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 20      chr16  64387109  64387209      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr16&#39;, &#39;-&#39;)
## [   Chromosome     Start       End Strand
## 24      chr16  58211024  58211124      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 9       chr16  32763157  32763257      -
## 10      chr16  71111412  71111512      -]
## (&#39;chr17&#39;, &#39;-&#39;)
## [   Chromosome     Start       End Strand
## 17      chr17  54630874  54630974      -,    Chromosome     Start       End Strand
## 19      chr17  14894477  14894577      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr18&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 0      chr18  15185403  15185503      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr19&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 22      chr19  34184458  34184558      +
## 23      chr19  20380555  20380655      +,    Chromosome     Start       End Strand
## 23      chr19  47351212  47351312      +]
## (&#39;chr19&#39;, &#39;-&#39;)
## [   Chromosome     Start       End Strand
## 16      chr19  22826941  22827041      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr20&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome    Start      End Strand
## 20      chr20  3725276  3725376      -]
## (&#39;chrX&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome     Start       End Strand
## 21       chrX  67819564  67819664      +,    Chromosome    Start      End Strand
## 11       chrX  7400690  7400790      +]
## (&#39;chrX&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],    Chromosome      Start        End Strand
## 12       chrX  147019154  147019254      -]
## (&#39;chrY&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 7       chrY  50096401  50096501      +,   Chromosome     Start       End Strand
## 0       chrY  13511383  13511483      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chrY&#39;, &#39;-&#39;)
## [  Chromosome     Start       End Strand
## 6       chrY  23164860  23164960      -,   Chromosome     Start       End Strand
## 1       chrY  23751710  23751810      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]</code></pre>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="printing-pyranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="sorting-pyranges.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
