
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>pyranges.readers &#8212; pyranges  documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-pyranges.readers">
<span id="pyranges-readers"></span><h1><a class="reference internal" href="#module-pyranges.readers" title="pyranges.readers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.readers</span></code></a><a class="headerlink" href="#module-pyranges.readers" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pyranges.readers.read_bed">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">read_bed</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">as_df=False</em>, <em class="sig-param">nrows=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.read_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Return bed file as PyRanges.</p>
<p>This is a reader for files that follow the bed format. They can have from
3-12 columns which will be named like so:</p>
<p>Chromosome Start End Name Score Strand ThickStart ThickEnd ItemRGB
BlockCount BlockSizes BlockStarts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to bed file</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>nrows</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of rows to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If you just want to create a PyRanges from a tab-delimited bed-like file,
use <cite>pr.PyRanges(pandas.read_table(f))</cite> instead.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_example_path</span><span class="p">(</span><span class="s2">&quot;aorta.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |      9939 |     10138 | H3K27me3   |         7 | +            |</span>
<span class="go">| chr1         |      9953 |     10152 | H3K27me3   |         5 | +            |</span>
<span class="go">| chr1         |      9916 |     10115 | H3K27me3   |         5 | -            |</span>
<span class="go">| chr1         |      9951 |     10150 | H3K27me3   |         8 | -            |</span>
<span class="go">| chr1         |      9978 |     10177 | H3K27me3   |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">as_df</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">  Chromosome  Start    End      Name  Score Strand</span>
<span class="go">0       chr1   9916  10115  H3K27me3      5      -</span>
<span class="go">1       chr1   9939  10138  H3K27me3      7      +</span>
<span class="go">2       chr1   9951  10150  H3K27me3      8      -</span>
<span class="go">3       chr1   9953  10152  H3K27me3      5      +</span>
<span class="go">4       chr1   9978  10177  H3K27me3      7      -</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.readers.read_bam">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">read_bam</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">sparse=True</em>, <em class="sig-param">as_df=False</em>, <em class="sig-param">mapq=0</em>, <em class="sig-param">required_flag=0</em>, <em class="sig-param">filter_flag=1540</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.read_bam" title="Permalink to this definition">¶</a></dt>
<dd><p>Return bam file as PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to bam file</p></li>
<li><p><strong>sparse</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to return only.</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>mapq</strong> (<em>int</em><em>, </em><em>default 0</em>) – Minimum mapping quality score.</p></li>
<li><p><strong>required_flag</strong> (<em>int</em><em>, </em><em>default 0</em>) – Flags which must be present for the interval to be read.</p></li>
<li><p><strong>filter_flag</strong> (<em>int</em><em>, </em><em>default 1540</em>) – Ignore reads with these flags. Default 1540, which means that either
the read is unmapped, the read failed vendor or platfrom quality
checks, or the read is a PCR or optical duplicate.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This functionality requires the library <cite>bamread</cite>. It can be installed with
<cite>pip install bamread</cite> or <cite>conda install -c bioconda bamread</cite>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_example_path</span><span class="p">(</span><span class="s2">&quot;control.bam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | Flag       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (uint16)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         | 887771    | 887796    | +            | 16         |</span>
<span class="go">| chr1         | 994660    | 994685    | +            | 16         |</span>
<span class="go">| chr1         | 1770383   | 1770408   | +            | 16         |</span>
<span class="go">| chr1         | 1995141   | 1995166   | +            | 16         |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...        |</span>
<span class="go">| chrY         | 57402214  | 57402239  | +            | 16         |</span>
<span class="go">| chrY         | 10643526  | 10643551  | -            | 0          |</span>
<span class="go">| chrY         | 11776321  | 11776346  | -            | 0          |</span>
<span class="go">| chrY         | 20557165  | 20557190  | -            | 0          |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 5 columns from 25 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">sparse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">as_df</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">     Chromosome    Start      End Strand  Flag  QueryStart  QueryEnd Name Cigar Quality</span>
<span class="go">0          chr1   887771   887796      +    16           0        25   U0   25M    None</span>
<span class="go">1          chr1   994660   994685      +    16           0        25   U0   25M    None</span>
<span class="go">2          chr1  1041102  1041127      -     0           0        25   U0   25M    None</span>
<span class="go">3          chr1  1770383  1770408      +    16           0        25   U0   25M    None</span>
<span class="go">4          chr1  1995141  1995166      +    16           0        25   U0   25M    None</span>
<span class="gp">... </span>        <span class="o">...</span>      <span class="o">...</span>      <span class="o">...</span>    <span class="o">...</span>   <span class="o">...</span>         <span class="o">...</span>       <span class="o">...</span>  <span class="o">...</span>   <span class="o">...</span>     <span class="o">...</span>
<span class="go">9995       chrM     3654     3679      -     0           0        25   U0   25M    None</span>
<span class="go">9996       chrM     3900     3925      +    16           0        25   U0   25M    None</span>
<span class="go">9997       chrM    13006    13031      +    16           0        25   U0   25M    None</span>
<span class="go">9998       chrM    14257    14282      -     0           0        25   U0   25M    None</span>
<span class="go">9999       chrM    14257    14282      -     0           0        25   U0   25M    None</span>

<span class="go">[10000 rows x 10 columns]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.readers._fetch_gene_transcript_exon_id">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">_fetch_gene_transcript_exon_id</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">annotation=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers._fetch_gene_transcript_exon_id" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyranges.readers.skiprows">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">skiprows</code><span class="sig-paren">(</span><em class="sig-param">f</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.skiprows" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyranges.readers.read_gtf">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">read_gtf</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">full=True</em>, <em class="sig-param">as_df=False</em>, <em class="sig-param">nrows=None</em>, <em class="sig-param">duplicate_attr=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.read_gtf" title="Permalink to this definition">¶</a></dt>
<dd><p>Read files in the Gene Transfer Format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to GTF file.</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>nrows</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of rows to read. Default None, i.e. all.</p></li>
<li><p><strong>duplicate_attr</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to handle (potential) duplicate attributes or just keep last one.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_gff3" title="pyranges.read_gff3"><code class="xref py py-func docutils literal notranslate"><span class="pre">pyranges.read_gff3()</span></code></a></dt><dd><p>read files in the General Feature Format</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_example_path</span><span class="p">(</span><span class="s2">&quot;ensembl.gtf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">read_gtf</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># +--------------+------------+--------------+-----------+-----------+------------+--------------+------------+-----------------+----------------+-------+</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id         | gene_version   | +18   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)        | (object)       | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># |--------------+------------+--------------+-----------+-----------+------------+--------------+------------+-----------------+----------------+-------|</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | gene         | 11868     | 14409     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | transcript   | 11868     | 14409     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | exon         | 11868     | 12227     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | exon         | 12612     | 12721     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...             | ...            | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | transcript   | 120724    | 133723    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | exon         | 133373    | 133723    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | exon         | 129054    | 129223    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | exon         | 120873    | 120932    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># +--------------+------------+--------------+-----------+-----------+------------+--------------+------------+-----------------+----------------+-------+</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Stranded PyRanges object has 95 rows and 28 columns from 1 chromosomes.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># 18 hidden columns: gene_name, gene_source, gene_biotype, transcript_id, transcript_version, transcript_name, transcript_source, transcript_biotype, tag, transcript_support_level, ... (+ 8 more.)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.readers.read_gtf_full">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">read_gtf_full</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">as_df=False</em>, <em class="sig-param">nrows=None</em>, <em class="sig-param">skiprows=0</em>, <em class="sig-param">duplicate_attr=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.read_gtf_full" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyranges.readers.to_rows">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">to_rows</code><span class="sig-paren">(</span><em class="sig-param">anno</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.to_rows" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyranges.readers.to_rows_keep_duplicates">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">to_rows_keep_duplicates</code><span class="sig-paren">(</span><em class="sig-param">anno</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.to_rows_keep_duplicates" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyranges.readers.read_gtf_restricted">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">read_gtf_restricted</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">as_df=False</em>, <em class="sig-param">skiprows=0</em>, <em class="sig-param">nrows=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.read_gtf_restricted" title="Permalink to this definition">¶</a></dt>
<dd><p>seqname - name of the chromosome or scaffold; chromosome names can be given with or without the ‘chr’ prefix. Important note: the seqname must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.
# source - name of the program that generated this feature, or the data source (database or project name)
feature - feature type name, e.g. Gene, Variation, Similarity
start - Start position of the feature, with sequence numbering starting at 1.
end - End position of the feature, with sequence numbering starting at 1.
score - A floating point value.
strand - defined as + (forward) or - (reverse).
# frame - One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on..
attribute - A semicolon-separated list of tag-value pairs, providing additional information about each feature.</p>
</dd></dl>

<dl class="function">
<dt id="pyranges.readers.to_rows_gff3">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">to_rows_gff3</code><span class="sig-paren">(</span><em class="sig-param">anno</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.to_rows_gff3" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="pyranges.readers.read_gff3">
<code class="sig-prename descclassname">pyranges.readers.</code><code class="sig-name descname">read_gff3</code><span class="sig-paren">(</span><em class="sig-param">f</em>, <em class="sig-param">annotation=None</em>, <em class="sig-param">as_df=False</em>, <em class="sig-param">nrows=None</em>, <em class="sig-param">skiprows=0</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.readers.read_gff3" title="Permalink to this definition">¶</a></dt>
<dd><p>Read files in the General Feature Format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to GFF file.</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>nrows</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of rows to read. Default None, i.e. all.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_gtf" title="pyranges.read_gtf"><code class="xref py py-func docutils literal notranslate"><span class="pre">pyranges.read_gtf()</span></code></a></dt><dd><p>read files in the Gene Transfer Format</p>
</dd>
</dl>
</div>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">pyranges</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">API Reference</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Endre Bakken Stovner.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../../_sources/autoapi/pyranges/readers/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>