
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>pyranges.statistics &#8212; pyranges  documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-pyranges.statistics">
<span id="pyranges-statistics"></span><h1><a class="reference internal" href="#module-pyranges.statistics" title="pyranges.statistics"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.statistics</span></code></a><a class="headerlink" href="#module-pyranges.statistics" title="Permalink to this headline">¶</a></h1>
<p><p>Statistics useful for genomics.</p>
</p>
<div class="section" id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="pyranges.statistics.fdr">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">fdr</code><span class="sig-paren">(</span><em class="sig-param">p_vals</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.fdr" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust p-values with Benjamini-Hochberg.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>array-like</em>) – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame where values are order of data.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span> <span class="o">/</span> <span class="mi">100</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">PValue</span> <span class="o">=</span> <span class="n">x</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+----------------------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | PValue               |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (float64)            |</span>
<span class="go">|--------------+-----------+-----------+--------------+----------------------|</span>
<span class="go">| chr1         | 176601938 | 176602038 | +            | 0.005488135039273248 |</span>
<span class="go">| chr1         | 155082851 | 155082951 | -            | 0.007151893663724195 |</span>
<span class="go">| chr2         | 211134424 | 211134524 | -            | 0.006027633760716439 |</span>
<span class="go">| chr9         | 78826761  | 78826861  | -            | 0.005448831829968969 |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...                  |</span>
<span class="go">| chr16        | 52216522  | 52216622  | +            | 0.004375872112626925 |</span>
<span class="go">| chr17        | 8085927   | 8086027   | -            | 0.008917730007820798 |</span>
<span class="go">| chr19        | 17333425  | 17333525  | +            | 0.009636627605010294 |</span>
<span class="go">| chr22        | 16728001  | 16728101  | +            | 0.003834415188257777 |</span>
<span class="go">+--------------+-----------+-----------+--------------+----------------------+</span>
<span class="go">Stranded PyRanges object has 10 rows and 5 columns from 9 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">FDR</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">fdr</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">PValue</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">print</span><span class="p">(</span><span class="n">formatting</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;PValue&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{:.4f}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;FDR&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{:.4}</span><span class="s2">&quot;</span><span class="p">})</span>
<span class="go">+--------------+-----------+-----------+--------------+-------------+-------------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | PValue      | FDR         |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (float64)   | (float64)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+-------------+-------------|</span>
<span class="go">| chr1         | 176601938 | 176602038 | +            | 0.0055      | 0.01098     |</span>
<span class="go">| chr1         | 155082851 | 155082951 | -            | 0.0072      | 0.00894     |</span>
<span class="go">| chr2         | 211134424 | 211134524 | -            | 0.0060      | 0.01005     |</span>
<span class="go">| chr9         | 78826761  | 78826861  | -            | 0.0054      | 0.01362     |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...         | ...         |</span>
<span class="go">| chr16        | 52216522  | 52216622  | +            | 0.0044      | 0.01459     |</span>
<span class="go">| chr17        | 8085927   | 8086027   | -            | 0.0089      | 0.009909    |</span>
<span class="go">| chr19        | 17333425  | 17333525  | +            | 0.0096      | 0.009637    |</span>
<span class="go">| chr22        | 16728001  | 16728101  | +            | 0.0038      | 0.03834     |</span>
<span class="go">+--------------+-----------+-----------+--------------+-------------+-------------+</span>
<span class="go">Stranded PyRanges object has 10 rows and 6 columns from 9 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.statistics.fisher_exact">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">fisher_exact</code><span class="sig-paren">(</span><em class="sig-param">n1</em>, <em class="sig-param">d1</em>, <em class="sig-param">n2</em>, <em class="sig-param">d2</em>, <em class="sig-param">pseudocount=0</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.fisher_exact" title="Permalink to this definition">¶</a></dt>
<dd><p>Fisher’s exact for contingency tables.</p>
<p>Computes the hypotheses two-sided, less and greater at the same time.</p>
<p>The odds-ratio is</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>array-like of int</em>) – Top left square of contingency table.</p></li>
<li><p><strong>d1</strong> (<em>array-like of int</em>) – Bottom left square of contingency table.</p></li>
<li><p><strong>n2</strong> (<em>array-like of int</em>) – Top right square of contingency table.</p></li>
<li><p><strong>d2</strong> (<em>array-like of int</em>) – Bottom right square of contingency table.</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em><em>, </em><em>default 0</em>) – Values &gt; 0 allow Odds Ratio to always be a finite number.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The odds-ratio is computed thusly:</p>
<p><code class="docutils literal notranslate"><span class="pre">((n1</span> <span class="pre">+</span> <span class="pre">pseudocount)</span> <span class="pre">/</span> <span class="pre">(d2</span> <span class="pre">+</span> <span class="pre">pseudocount))</span> <span class="pre">/</span> <span class="pre">((n2</span> <span class="pre">+</span> <span class="pre">pseudocount)</span> <span class="pre">/</span> <span class="pre">(d1</span> <span class="pre">+</span> <span class="pre">pseudocount))</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>DataFrame with columns OR and P, PLeft and PRight.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-func docutils literal notranslate"><span class="pre">pr.stats.fdr()</span></code></dt><dd><p>correct for multiple testing</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;TP&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="s2">&quot;FP&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">11</span><span class="p">,</span> <span class="mi">12</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;TN&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;FN&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span>
<span class="go">   TP  FP  TN  FN</span>
<span class="go">0   1  11   9   3</span>
<span class="go">1   0  12  10   2</span>
<span class="go">2   8   1   2   5</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">fisher_exact</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">TP</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">FP</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">TN</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">FN</span><span class="p">)</span>
<span class="go">         OR         P     PLeft    PRight</span>
<span class="go">0  0.407407  0.002759  0.001380  0.999966</span>
<span class="go">1  0.000000  0.000067  0.000034  1.000000</span>
<span class="go">2  0.800000  0.034965  0.999126  0.024476</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.statistics.mcc">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">mcc</code><span class="sig-paren">(</span><em class="sig-param">grs</em>, <em class="sig-param">genome=None</em>, <em class="sig-param">labels=None</em>, <em class="sig-param">strand=False</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.mcc" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Matthew’s correlation coefficient for PyRanges overlaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>grs</strong> (<em>list of PyRanges</em>) – PyRanges to compare.</p></li>
<li><p><strong>genome</strong> (<em>DataFrame</em><em> or </em><em>dict</em><em>, </em><em>default None</em>) – Should contain chromosome sizes. By default, end position of the
rightmost intervals are used as proxies for the chromosome size, but
it is recommended to use a genome.</p></li>
<li><p><strong>labels</strong> (<em>list of str</em><em>, </em><em>default None</em>) – Names to give the PyRanges in the output.</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to compute correlations per strand.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>default False</em>) – Warn if some chromosomes are in the genome, but not in the PyRanges.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;chrM&quot;</span><span class="p">:</span> <span class="mi">16000</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">grs</span> <span class="o">=</span> <span class="p">[</span><span class="n">pr</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">chromsizes</span><span class="o">=</span><span class="n">chromsizes</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mcc</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">mcc</span><span class="p">(</span><span class="n">grs</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="n">chromsizes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mcc</span>
<span class="go">   T  F     TP  FP  TN  FN       MCC</span>
<span class="go">0  a  a  15920   0  80   0  1.000000</span>
<span class="go">1  a  b  15875  65  15  45  0.213109</span>
<span class="go">3  a  c  15896  72   8  24  0.155496</span>
<span class="go">2  b  a  15875  45  15  65  0.213109</span>
<span class="go">5  b  b  15940   0  60   0  1.000000</span>
<span class="go">6  b  c  15916  52   8  24  0.180354</span>
<span class="go">4  c  a  15896  24   8  72  0.155496</span>
<span class="go">7  c  b  15916  24   8  52  0.180354</span>
<span class="go">8  c  c  15968   0  32   0  1.000000</span>
</pre></div>
</div>
<p>To create a symmetric matrix (useful for heatmaps of correlations):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mcc</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;F&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">MCC</span><span class="o">.</span><span class="n">unstack</span><span class="p">()</span>
<span class="go">F         a         b         c</span>
<span class="go">T</span>
<span class="go">a  1.000000  0.213109  0.155496</span>
<span class="go">b  0.213109  1.000000  0.180354</span>
<span class="go">c  0.155496  0.180354  1.000000</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.statistics.rowbased_spearman">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">rowbased_spearman</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">y</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.rowbased_spearman" title="Permalink to this definition">¶</a></dt>
<dd><p>Fast row-based Spearman’s correlation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>matrix-like</em>) – 2D numerical matrix. Same size as y.</p></li>
<li><p><strong>y</strong> (<em>matrix-like</em>) – 2D numerical matrix. Same size as x.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array with same length as input, where values are P-values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.array</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.statistics.rowbased_pearson" title="pyranges.statistics.rowbased_pearson"><code class="xref py py-func docutils literal notranslate"><span class="pre">pyranges.statistics.rowbased_pearson()</span></code></a></dt><dd><p>fast row-based Pearson’s correlation.</p>
</dd>
<dt><code class="xref py py-func docutils literal notranslate"><span class="pre">pr.stats.fdr()</span></code></dt><dd><p>correct for multiple testing</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
<p>Perform Spearman’s correlation pairwise on each row in 10x10 matrixes:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">rowbased_spearman</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
<span class="go">array([ 0.07523548, -0.24838724,  0.03703774,  0.24194052,  0.04778621,</span>
<span class="go">       -0.23913505,  0.12923138,  0.26840486,  0.13292204, -0.29846295])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.statistics.rowbased_pearson">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">rowbased_pearson</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">y</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.rowbased_pearson" title="Permalink to this definition">¶</a></dt>
<dd><p>Fast row-based Pearson’s correlation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>matrix-like</em>) – 2D numerical matrix. Same size as y.</p></li>
<li><p><strong>y</strong> (<em>matrix-like</em>) – 2D numerical matrix. Same size as x.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array with same length as input, where values are P-values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.array</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.statistics.rowbased_spearman" title="pyranges.statistics.rowbased_spearman"><code class="xref py py-func docutils literal notranslate"><span class="pre">pyranges.statistics.rowbased_spearman()</span></code></a></dt><dd><p>fast row-based Spearman’s correlation.</p>
</dd>
<dt><code class="xref py py-func docutils literal notranslate"><span class="pre">pr.stats.fdr()</span></code></dt><dd><p>correct for multiple testing</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
<p>Perform Pearson’s correlation pairwise on each row in 10x10 matrixes:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">rowbased_pearson</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
<span class="go">array([ 0.20349603, -0.01667236, -0.01448763, -0.00442322,  0.06527234,</span>
<span class="go">       -0.36710862,  0.14978726,  0.32360286,  0.17209191, -0.08902829])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.statistics.rowbased_rankdata">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">rowbased_rankdata</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.rowbased_rankdata" title="Permalink to this definition">¶</a></dt>
<dd><p>Rank order of entries in each row.</p>
<p>Same as SciPy rankdata with method=mean.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>matrix-like</em>) – The data to find order of.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame where values are order of data.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[5, 0, 3, 3, 7, 9, 3, 5, 2, 4],</span>
<span class="go">       [7, 6, 8, 8, 1, 6, 7, 7, 8, 1],</span>
<span class="go">       [5, 9, 8, 9, 4, 3, 0, 3, 5, 0]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">rowbased_rankdata</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="go">     0    1    2    3    4     5    6    7    8    9</span>
<span class="go">0  7.5  1.0  4.0  4.0  9.0  10.0  4.0  7.5  2.0  6.0</span>
<span class="go">1  6.0  3.5  9.0  9.0  1.5   3.5  6.0  6.0  9.0  1.5</span>
<span class="go">2  6.5  9.5  8.0  9.5  5.0   3.5  1.5  3.5  6.5  1.5</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.statistics.simes">
<code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">simes</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">groupby</em>, <em class="sig-param">pcol</em>, <em class="sig-param">keep_position=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.simes" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply Simes method for giving dependent events a p-value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data to analyse with Simes.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em> or </em><em>list of str</em>) – Features equal in these columns will be merged with Simes.</p></li>
<li><p><strong>pcol</strong> (<em>str</em>) – Name of column with p-values.</p></li>
<li><p><strong>keep_position</strong> (<em>bool</em><em>, </em><em>default False</em>) – Keep columns “Chromosome”, “Start”, “End” and “Strand” if they exist.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-func docutils literal notranslate"><span class="pre">pr.stats.fdr()</span></code></dt><dd><p>correct for multiple testing</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;&#39;&#39;Chromosome Start End Strand Gene PValue</span>
<span class="gp">... </span><span class="s1">1 10 20 + P53 0.0001</span>
<span class="gp">... </span><span class="s1">1 20 20 + P53 0.0002</span>
<span class="gp">... </span><span class="s1">1 30 20 + P53 0.0003</span>
<span class="gp">... </span><span class="s1">2 60 65 - FOX 0.05</span>
<span class="gp">... </span><span class="s1">2 70 75 - FOX 0.0000001</span>
<span class="gp">... </span><span class="s1">2 80 90 - FOX 0.0000021&#39;&#39;&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+-------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       | Gene       |      PValue |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   | (object)   |   (float64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------+-------------|</span>
<span class="go">|            1 |        10 |        20 | +            | P53        |     0.0001  |</span>
<span class="go">|            1 |        20 |        20 | +            | P53        |     0.0002  |</span>
<span class="go">|            1 |        30 |        20 | +            | P53        |     0.0003  |</span>
<span class="go">|            2 |        60 |        65 | -            | FOX        |     0.05    |</span>
<span class="go">|            2 |        70 |        75 | -            | FOX        |     1e-07   |</span>
<span class="go">|            2 |        80 |        90 | -            | FOX        |     2.1e-06 |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+-------------+</span>
<span class="go">Stranded PyRanges object has 6 rows and 6 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">simes</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">simes</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;PValue&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simes</span>
<span class="go">  Gene         Simes</span>
<span class="go">0  FOX  3.000000e-07</span>
<span class="go">1  P53  3.000000e-04</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span>
<span class="gp">... </span><span class="n">pr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">simes</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;PValue&quot;</span><span class="p">,</span> <span class="n">keep_position</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="go">+--------------+-----------+-----------+-------------+------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |       Simes | Strand     | Gene       |</span>
<span class="go">|     (object) |   (int32) |   (int32) |   (float64) | (object)   | (object)   |</span>
<span class="go">|--------------+-----------+-----------+-------------+------------+------------|</span>
<span class="go">|            1 |        10 |        20 |      0.0001 | +          | P53        |</span>
<span class="go">|            2 |        60 |        90 |      1e-07  | -          | FOX        |</span>
<span class="go">+--------------+-----------+-----------+-------------+------------+------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="pyranges.statistics.StatisticsMethods">
<em class="property">class </em><code class="sig-prename descclassname">pyranges.statistics.</code><code class="sig-name descname">StatisticsMethods</code><span class="sig-paren">(</span><em class="sig-param">pr</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.StatisticsMethods" title="Permalink to this definition">¶</a></dt>
<dd><p>Namespace for statistical comparsion-operations.</p>
<p>Accessed with gr.stats.</p>
<dl class="attribute">
<dt id="pyranges.statistics.StatisticsMethods.pr">
<code class="sig-name descname">pr</code><a class="headerlink" href="#pyranges.statistics.StatisticsMethods.pr" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pyranges.statistics.StatisticsMethods.forbes">
<code class="sig-name descname">forbes</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em>, <em class="sig-param">chromsizes</em>, <em class="sig-param">strandedness=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.StatisticsMethods.forbes" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Forbes coefficient.</p>
<p>Ratio which represents observed versus expected co-occurence.</p>
<p>Described in <code class="docutils literal notranslate"><span class="pre">Forbes</span> <span class="pre">SA</span> <span class="pre">(1907):</span> <span class="pre">On</span> <span class="pre">the</span> <span class="pre">local</span> <span class="pre">distribution</span> <span class="pre">of</span> <span class="pre">certain</span> <span class="pre">Illinois</span> <span class="pre">fishes:</span> <span class="pre">an</span> <span class="pre">essay</span> <span class="pre">in</span> <span class="pre">statistical</span> <span class="pre">ecology.</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Intervals to compare with.</p></li>
<li><p><strong>chromsizes</strong> (<em>int</em><em>, </em><em>dict</em><em>, </em><em>DataFrame</em><em> or </em><a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Integer representing genome length or mapping from chromosomes
to its length.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. &quot;auto&quot;</em>) – Whether to compute without regards to strand or on same or opposite.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Ratio of observed versus expected co-occurence.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.statistics.jaccard()</span></code></dt><dd><p>compute the jaccard coefficient</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="p">,</span> <span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">(),</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq_background</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">forbes</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">chromsizes</span><span class="o">=</span><span class="n">chromsizes</span><span class="p">)</span>
<span class="go">1.7168314674978278</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pyranges.statistics.StatisticsMethods.jaccard">
<code class="sig-name descname">jaccard</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.StatisticsMethods.jaccard" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Jaccards coefficient.</p>
<p>Ratio of the intersection and union of two sets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Intervals to compare with.</p></li>
<li><p><strong>chromsizes</strong> (<em>int</em><em>, </em><em>dict</em><em>, </em><em>DataFrame</em><em> or </em><a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Integer representing genome length or mapping from chromosomes
to its length.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. &quot;auto&quot;</em>) – Whether to compute without regards to strand or on same or opposite.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Ratio of the intersection and union of two sets.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.statistics.forbes()</span></code></dt><dd><p>compute the forbes coefficient</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="p">,</span> <span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">(),</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq_background</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">jaccard</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">chromsizes</span><span class="o">=</span><span class="n">chromsizes</span><span class="p">)</span>
<span class="go">6.657941988519211e-05</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pyranges.statistics.StatisticsMethods.relative_distance">
<code class="sig-name descname">relative_distance</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.statistics.StatisticsMethods.relative_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute spatial correllation between two sets.</p>
<p>Metric which describes relative distance between each interval in one
set and two closest intervals in another.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Intervals to compare with.</p></li>
<li><p><strong>chromsizes</strong> (<em>int</em><em>, </em><em>dict</em><em>, </em><em>DataFrame</em><em> or </em><a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Integer representing genome length or mapping from chromosomes
to its length.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. &quot;auto&quot;</em>) – Whether to compute without regards to strand or on same or opposite.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame containing the frequency of each relative distance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.statistics.jaccard()</span></code></dt><dd><p>compute the jaccard coefficient</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.statistics.forbes()</span></code></dt><dd><p>compute the forbes coefficient</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="p">,</span> <span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">(),</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq_background</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">relative_distance</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="go">    reldist  count  total  fraction</span>
<span class="go">0      0.00    264   9956  0.026517</span>
<span class="go">1      0.01    226   9956  0.022700</span>
<span class="go">2      0.02    206   9956  0.020691</span>
<span class="go">3      0.03    235   9956  0.023604</span>
<span class="go">4      0.04    194   9956  0.019486</span>
<span class="go">5      0.05    241   9956  0.024207</span>
<span class="go">6      0.06    201   9956  0.020189</span>
<span class="go">7      0.07    191   9956  0.019184</span>
<span class="go">8      0.08    192   9956  0.019285</span>
<span class="go">9      0.09    191   9956  0.019184</span>
<span class="go">10     0.10    186   9956  0.018682</span>
<span class="go">11     0.11    203   9956  0.020390</span>
<span class="go">12     0.12    218   9956  0.021896</span>
<span class="go">13     0.13    209   9956  0.020992</span>
<span class="go">14     0.14    201   9956  0.020189</span>
<span class="go">15     0.15    178   9956  0.017879</span>
<span class="go">16     0.16    202   9956  0.020289</span>
<span class="go">17     0.17    197   9956  0.019787</span>
<span class="go">18     0.18    208   9956  0.020892</span>
<span class="go">19     0.19    202   9956  0.020289</span>
<span class="go">20     0.20    191   9956  0.019184</span>
<span class="go">21     0.21    188   9956  0.018883</span>
<span class="go">22     0.22    213   9956  0.021394</span>
<span class="go">23     0.23    192   9956  0.019285</span>
<span class="go">24     0.24    199   9956  0.019988</span>
<span class="go">25     0.25    181   9956  0.018180</span>
<span class="go">26     0.26    172   9956  0.017276</span>
<span class="go">27     0.27    191   9956  0.019184</span>
<span class="go">28     0.28    190   9956  0.019084</span>
<span class="go">29     0.29    192   9956  0.019285</span>
<span class="go">30     0.30    201   9956  0.020189</span>
<span class="go">31     0.31    212   9956  0.021294</span>
<span class="go">32     0.32    213   9956  0.021394</span>
<span class="go">33     0.33    177   9956  0.017778</span>
<span class="go">34     0.34    197   9956  0.019787</span>
<span class="go">35     0.35    163   9956  0.016372</span>
<span class="go">36     0.36    191   9956  0.019184</span>
<span class="go">37     0.37    198   9956  0.019888</span>
<span class="go">38     0.38    160   9956  0.016071</span>
<span class="go">39     0.39    188   9956  0.018883</span>
<span class="go">40     0.40    200   9956  0.020088</span>
<span class="go">41     0.41    188   9956  0.018883</span>
<span class="go">42     0.42    230   9956  0.023102</span>
<span class="go">43     0.43    197   9956  0.019787</span>
<span class="go">44     0.44    224   9956  0.022499</span>
<span class="go">45     0.45    184   9956  0.018481</span>
<span class="go">46     0.46    198   9956  0.019888</span>
<span class="go">47     0.47    187   9956  0.018783</span>
<span class="go">48     0.48    200   9956  0.020088</span>
<span class="go">49     0.49    194   9956  0.019486</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">pyranges</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">API Reference</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Endre Bakken Stovner.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../../_sources/autoapi/pyranges/statistics/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>