
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>pyranges.genomicfeatures &#8212; pyranges  documentation</title>
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-pyranges.genomicfeatures">
<span id="pyranges-genomicfeatures"></span><h1><a class="reference internal" href="#module-pyranges.genomicfeatures" title="pyranges.genomicfeatures"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.genomicfeatures</span></code></a><a class="headerlink" href="#module-pyranges.genomicfeatures" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pyranges.genomicfeatures.GenomicFeaturesMethods">
<em class="property">class </em><code class="sig-prename descclassname">pyranges.genomicfeatures.</code><code class="sig-name descname">GenomicFeaturesMethods</code><span class="sig-paren">(</span><em class="sig-param">pr</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.genomicfeatures.GenomicFeaturesMethods" title="Permalink to this definition">¶</a></dt>
<dd><p>Namespace for methods using feature information.</p>
<p>Accessed through <cite>gr.features</cite>.</p>
<dl class="attribute">
<dt id="pyranges.genomicfeatures.GenomicFeaturesMethods.pr">
<code class="sig-name descname">pr</code><a class="headerlink" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.pr" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pyranges.genomicfeatures.GenomicFeaturesMethods.tss">
<code class="sig-name descname">tss</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.tss" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the transcription start sites.</p>
<p>Returns the 5’ for every interval with feature “transcript”.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.tes" title="pyranges.genomicfeatures.GenomicFeaturesMethods.tes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.genomicfeatures.GenomicFeaturesMethods.tes()</span></code></a></dt><dd><p>return the transcription end sites</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_biotype                       | +19   |</span>
<span class="go">| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                           | ...   |</span>
<span class="go">|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------|</span>
<span class="go">| 1            | havana     | gene         | 11868     | 14409     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | transcript   | 11868     | 14409     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | exon         | 11868     | 12227     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | exon         | 12612     | 12721     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...                                | ...   |</span>
<span class="go">| 1            | havana     | gene         | 1173055   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | transcript   | 1173055   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | exon         | 1179364   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | exon         | 1173055   | 1176396   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">19 hidden columns: gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, transcript_source, transcript_support_level, ... (+ 9 more.)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">tss</span><span class="p">()</span>
<span class="go">+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_biotype                       | +19   |</span>
<span class="go">| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                           | ...   |</span>
<span class="go">|--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------|</span>
<span class="go">| 1            | havana     | tss        | 11868     | 11869     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | tss        | 12009     | 12010     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | tss        | 29553     | 29554     | .          | +            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | tss        | 30266     | 30267     | .          | +            | .          | lncRNA                             | ...   |</span>
<span class="go">| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                                | ...   |</span>
<span class="go">| 1            | havana     | tss        | 1092813   | 1092814   | .          | -            | .          | protein_coding                     | ...   |</span>
<span class="go">| 1            | havana     | tss        | 1116087   | 1116088   | .          | -            | .          | protein_coding                     | ...   |</span>
<span class="go">| 1            | havana     | tss        | 1116089   | 1116090   | .          | -            | .          | protein_coding                     | ...   |</span>
<span class="go">| 1            | havana     | tss        | 1179555   | 1179556   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">Stranded PyRanges object has 280 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">19 hidden columns: gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, transcript_source, transcript_support_level, ... (+ 9 more.)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pyranges.genomicfeatures.GenomicFeaturesMethods.tes">
<code class="sig-name descname">tes</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">slack=0</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.tes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the transcription end sites.</p>
<p>Returns the 3’ for every interval with feature “transcript”.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.tss" title="pyranges.genomicfeatures.GenomicFeaturesMethods.tss"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.genomicfeatures.GenomicFeaturesMethods.tss()</span></code></a></dt><dd><p>return the transcription start sites</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_biotype                       | +19   |</span>
<span class="go">| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                           | ...   |</span>
<span class="go">|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------|</span>
<span class="go">| 1            | havana     | gene         | 11868     | 14409     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | transcript   | 11868     | 14409     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | exon         | 11868     | 12227     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | exon         | 12612     | 12721     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...                                | ...   |</span>
<span class="go">| 1            | havana     | gene         | 1173055   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | transcript   | 1173055   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | exon         | 1179364   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | exon         | 1173055   | 1176396   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">19 hidden columns: gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, transcript_source, transcript_support_level, ... (+ 9 more.)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">tes</span><span class="p">()</span>
<span class="go">+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">| Chromosome   | Source     | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_biotype                       | +19   |</span>
<span class="go">| (category)   | (object)   | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                           | ...   |</span>
<span class="go">|--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------|</span>
<span class="go">| 1            | havana     | tes        | 14409     | 14410     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | tes        | 13670     | 13671     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | tes        | 31097     | 31098     | .          | +            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | tes        | 31109     | 31110     | .          | +            | .          | lncRNA                             | ...   |</span>
<span class="go">| ...          | ...        | ...        | ...       | ...       | ...        | ...          | ...        | ...                                | ...   |</span>
<span class="go">| 1            | havana     | tes        | 1092813   | 1092814   | .          | -            | .          | protein_coding                     | ...   |</span>
<span class="go">| 1            | havana     | tes        | 1116087   | 1116088   | .          | -            | .          | protein_coding                     | ...   |</span>
<span class="go">| 1            | havana     | tes        | 1116089   | 1116090   | .          | -            | .          | protein_coding                     | ...   |</span>
<span class="go">| 1            | havana     | tes        | 1179555   | 1179556   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">+--------------+------------+------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">Stranded PyRanges object has 280 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">19 hidden columns: gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, transcript_source, transcript_support_level, ... (+ 9 more.)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pyranges.genomicfeatures.GenomicFeaturesMethods.introns">
<code class="sig-name descname">introns</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">by='gene'</em>, <em class="sig-param">nb_cpu=1</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.introns" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the introns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>str</em><em>, </em><em>{&quot;gene&quot;</em><em>, </em><em>&quot;transcript&quot;}</em><em>, </em><em>default &quot;gene&quot;</em>) – Whether to find introns per gene or transcript.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.genomicfeatures.GenomicFeaturesMethods.tss" title="pyranges.genomicfeatures.GenomicFeaturesMethods.tss"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pyranges.genomicfeatures.GenomicFeaturesMethods.tss()</span></code></a></dt><dd><p>return the transcription start sites</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">| Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_biotype                       | +19   |</span>
<span class="go">| (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                           | ...   |</span>
<span class="go">|--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------|</span>
<span class="go">| 1            | havana     | gene         | 11868     | 14409     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | transcript   | 11868     | 14409     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | exon         | 11868     | 12227     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| 1            | havana     | exon         | 12612     | 12721     | .          | +            | .          | transcribed_unprocessed_pseudogene | ...   |</span>
<span class="go">| ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...                                | ...   |</span>
<span class="go">| 1            | havana     | gene         | 1173055   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | transcript   | 1173055   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | exon         | 1179364   | 1179555   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">| 1            | havana     | exon         | 1173055   | 1176396   | .          | -            | .          | lncRNA                             | ...   |</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+------------+--------------+------------+------------------------------------+-------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">19 hidden columns: gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, transcript_source, transcript_support_level, ... (+ 9 more.)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">introns</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
<span class="go">+--------------+----------------+------------+-----------+-----------+------------+--------------+------------+-------+</span>
<span class="go">| Chromosome   | Source         | Feature    | Start     | End       | Score      | Strand       | Frame      | +20   |</span>
<span class="go">| (object)     | (object)       | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | ...   |</span>
<span class="go">|--------------+----------------+------------+-----------+-----------+------------+--------------+------------+-------|</span>
<span class="go">| 1            | ensembl_havana | intron     | 1173926   | 1174265   | .          | +            | .          | ...   |</span>
<span class="go">| 1            | ensembl_havana | intron     | 1174321   | 1174423   | .          | +            | .          | ...   |</span>
<span class="go">| 1            | ensembl_havana | intron     | 1174489   | 1174520   | .          | +            | .          | ...   |</span>
<span class="go">| 1            | ensembl_havana | intron     | 1175034   | 1179188   | .          | +            | .          | ...   |</span>
<span class="go">| ...          | ...            | ...        | ...       | ...       | ...        | ...          | ...        | ...   |</span>
<span class="go">| 1            | havana         | intron     | 874591    | 875046    | .          | -            | .          | ...   |</span>
<span class="go">| 1            | havana         | intron     | 875155    | 875525    | .          | -            | .          | ...   |</span>
<span class="go">| 1            | havana         | intron     | 875625    | 876526    | .          | -            | .          | ...   |</span>
<span class="go">| 1            | havana         | intron     | 876611    | 876754    | .          | -            | .          | ...   |</span>
<span class="go">+--------------+----------------+------------+-----------+-----------+------------+--------------+------------+-------+</span>
<span class="go">Stranded PyRanges object has 311 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">20 hidden columns: gene_biotype, gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, ... (+ 11 more.)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">introns</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
<span class="go">+--------------+----------------+------------+-----------+-----------+------------+--------------+------------+----------------------------------+-------+</span>
<span class="go">| Chromosome   | Source         | Feature    | Start     | End       | Score      | Strand       | Frame      | gene_biotype                     | +19   |</span>
<span class="go">| (object)     | (object)       | (object)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)                         | ...   |</span>
<span class="go">|--------------+----------------+------------+-----------+-----------+------------+--------------+------------+----------------------------------+-------|</span>
<span class="go">| 1            | havana         | intron     | 818202    | 818722    | .          | +            | .          | lncRNA                           | ...   |</span>
<span class="go">| 1            | ensembl_havana | intron     | 960800    | 961292    | .          | +            | .          | protein_coding                   | ...   |</span>
<span class="go">| 1            | ensembl_havana | intron     | 961552    | 961628    | .          | +            | .          | protein_coding                   | ...   |</span>
<span class="go">| 1            | ensembl_havana | intron     | 961750    | 961825    | .          | +            | .          | protein_coding                   | ...   |</span>
<span class="go">| ...          | ...            | ...        | ...       | ...       | ...        | ...          | ...        | ...                              | ...   |</span>
<span class="go">| 1            | havana         | intron     | 732207    | 732980    | .          | -            | .          | transcribed_processed_pseudogene | ...   |</span>
<span class="go">| 1            | havana_tagene  | intron     | 168165    | 169048    | .          | -            | .          | lncRNA                           | ...   |</span>
<span class="go">| 1            | havana_tagene  | intron     | 165942    | 167958    | .          | -            | .          | lncRNA                           | ...   |</span>
<span class="go">| 1            | havana_tagene  | intron     | 168165    | 169048    | .          | -            | .          | lncRNA                           | ...   |</span>
<span class="go">+--------------+----------------+------------+-----------+-----------+------------+--------------+------------+----------------------------------+-------+</span>
<span class="go">Stranded PyRanges object has 1,043 rows and 28 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="go">19 hidden columns: gene_id, gene_name, gene_source, gene_version, tag, transcript_biotype, transcript_id, transcript_name, transcript_source, transcript_support_level, ... (+ 9 more.)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="pyranges.genomicfeatures.genome_bounds">
<code class="sig-prename descclassname">pyranges.genomicfeatures.</code><code class="sig-name descname">genome_bounds</code><span class="sig-paren">(</span><em class="sig-param">gr</em>, <em class="sig-param">chromsizes</em>, <em class="sig-param">clip=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.genomicfeatures.genome_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove or clip intervals outside of genome bounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromsizes</strong> (<em>dict</em><em> or </em><a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Dict or PyRanges describing the lengths of the chromosomes.</p></li>
<li><p><strong>clip</strong> (<em>bool</em><em>, </em><em>default False</em>) – Part of interval within bounds.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">249250600</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">249250640</span><span class="p">,</span> <span class="mi">7</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         2 |</span>
<span class="go">|            1 | 249250600 | 249250640 |</span>
<span class="go">|            3 |         5 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;1&quot;</span><span class="p">:</span> <span class="mi">249250621</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">:</span> <span class="mi">500</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span>
<span class="go">{&#39;1&#39;: 249250621, &#39;3&#39;: 500}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">gf</span><span class="o">.</span><span class="n">genome_bounds</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="n">chromsizes</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         2 |</span>
<span class="go">|            3 |         5 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">gf</span><span class="o">.</span><span class="n">genome_bounds</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="n">chromsizes</span><span class="p">,</span> <span class="n">clip</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         2 |</span>
<span class="go">|            1 | 249250600 | 249250621 |</span>
<span class="go">|            3 |         5 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">chromsizes</span><span class="p">[</span><span class="s1">&#39;3&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span>
<span class="go">{&#39;1&#39;: 249250621}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">gf</span><span class="o">.</span><span class="n">genome_bounds</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="n">chromsizes</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;3&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pyranges.genomicfeatures.tile_genome">
<code class="sig-prename descclassname">pyranges.genomicfeatures.</code><code class="sig-name descname">tile_genome</code><span class="sig-paren">(</span><em class="sig-param">genome</em>, <em class="sig-param">tile_size</em>, <em class="sig-param">tile_last=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.genomicfeatures.tile_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a tiled genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromsizes</strong> (<em>dict</em><em> or </em><a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Dict or PyRanges describing the lengths of the chromosomes.</p></li>
<li><p><strong>tile_size</strong> (<em>int</em>) – Length of the tiles.</p></li>
<li><p><strong>tile_last</strong> (<em>bool</em><em>, </em><em>default False</em>) – Use genome length as end of last tile.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.tile" title="pyranges.PyRanges.tile"><code class="xref py py-func docutils literal notranslate"><span class="pre">pyranges.PyRanges.tile()</span></code></a></dt><dd><p>split intervals into adjacent non-overlapping tiles.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromsizes</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   | Start     | End       |</span>
<span class="go">| (category)   | (int32)   | (int32)   |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         | 0         | 249250621 |</span>
<span class="go">| chr2         | 0         | 243199373 |</span>
<span class="go">| chr3         | 0         | 198022430 |</span>
<span class="go">| chr4         | 0         | 191154276 |</span>
<span class="go">| ...          | ...       | ...       |</span>
<span class="go">| chr22        | 0         | 51304566  |</span>
<span class="go">| chrM         | 0         | 16571     |</span>
<span class="go">| chrX         | 0         | 155270560 |</span>
<span class="go">| chrY         | 0         | 59373566  |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 25 rows and 3 columns from 25 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">gf</span><span class="o">.</span><span class="n">tile_genome</span><span class="p">(</span><span class="n">chromsizes</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mf">1e6</span><span class="p">))</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   | Start     | End       |</span>
<span class="go">| (category)   | (int32)   | (int32)   |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         | 0         | 1000000   |</span>
<span class="go">| chr1         | 1000000   | 2000000   |</span>
<span class="go">| chr1         | 2000000   | 3000000   |</span>
<span class="go">| chr1         | 3000000   | 4000000   |</span>
<span class="go">| ...          | ...       | ...       |</span>
<span class="go">| chrY         | 56000000  | 57000000  |</span>
<span class="go">| chrY         | 57000000  | 58000000  |</span>
<span class="go">| chrY         | 58000000  | 59000000  |</span>
<span class="go">| chrY         | 59000000  | 59373566  |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3,114 rows and 3 columns from 25 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">pyranges</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">API Reference</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Endre Bakken Stovner.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../../../_sources/autoapi/pyranges/genomicfeatures/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>